certstore_windows.go

```diff
@@ -386,22 +388,36 @@ func (wpk *winPrivateKey) cngSignHash(hash crypto.Hash, digest []byte) ([]byte,
 
 	// setup pkcs1v1.5 padding for RSA
 	if _, isRSA := wpk.publicKey.(*rsa.PublicKey); isRSA {
-		flags |= C.BCRYPT_PAD_PKCS1
-		padInfo := C.BCRYPT_PKCS1_PADDING_INFO{}
-		padPtr = unsafe.Pointer(&padInfo)
-
+		var algId C.LPCWSTR
 		switch hash {
 		case crypto.SHA1:
-			padInfo.pszAlgId = BCRYPT_SHA1_ALGORITHM
+			algId = BCRYPT_SHA1_ALGORITHM
 		case crypto.SHA256:
-			padInfo.pszAlgId = BCRYPT_SHA256_ALGORITHM
+			algId = BCRYPT_SHA256_ALGORITHM
 		case crypto.SHA384:
-			padInfo.pszAlgId = BCRYPT_SHA384_ALGORITHM
+			algId = BCRYPT_SHA384_ALGORITHM
 		case crypto.SHA512:
-			padInfo.pszAlgId = BCRYPT_SHA512_ALGORITHM
+			algId = BCRYPT_SHA512_ALGORITHM
 		default:
 			return nil, ErrUnsupportedHash
 		}
+
+		if pssOpts, ok := opts.(*rsa.PSSOptions); ok {
+			saltLen := pssOpts.SaltLength
+			if saltLen < 0 {
```

Comment:
`SaltLength` has two magic values, `PSSSaltLengthAuto` is 0 for an auto value, and `PSSSaltLengthEqualsHash` is -1.

We should not set the `saltLen` to `len(digest)` for anything other than PSSSaltLengthEqualsHash/-1 (maybe a new magic value will be introduced in the future).

I don't believe CNG will accept a value of "0" to mean "auto" the same way that it means in other places. [The documentation](https://github.com/golang/go/blob/01df2febf5f00b1dcba1843093ef99b338b23546/src/crypto/rsa/pss.go#L226-L227) states that auto is:

> PSSSaltLengthAuto causes the salt in a PSS signature to be as large as possible when signing, and to be auto-detected when verifying.

Can we replicate that auto length behavior if the value is 0?

Reply:
sure. nice catch :)